### ISSUE TITLE
Handling odd image sizes

### ISSUE THREAD

**Issue by farindk** [strukturag/libheif#191]
I often get asked why libheif does not save images with odd sizes.
The standard is unclear about images with odd size and I'd like to get other opinions about it.

The origin of the problem is that, as far as I can see, h265 does not allow odd image sizes at all for chroma 4:2:0. Every size (also the conformance window) is always specified in multiples of 2.

HEIF does not say anything about this. There is the 'ispe' box, which can contain
any value, but I am not sure whether it is a good idea to store there a different
value than the real h265 image size.

When encoding an image with odd size with libheif, it is currently rounded down to the nearest
even number. E.g. encoding 99x99 results in an image of size 98x98.
I know that this is not good, but I see no clean way to do it otherwise.
One option could be to save the image as 100x100, but store 99x99 in the 'ispe' box.

How do you think about this, I'd like to discuss this, before doing any change in this direction.

**Comment by farindk** [comment]
More thoughts:
The standard explicitly states that the 'ispe' box specifies the size of the 'reconstructed image', which is defined as the output of the decoding process. And since h265 only allows for even image sizes, I would say that this consequently also has to be an even size.

If we allowed odd sizes for 'ispe', we ran into all kinds of trouble when there are image transformations (rotation, cropping), as it is not clear anymore how the chroma samples relate to the luma samples.

**Comment by farindk** [comment]
@dlemstra @jonsneyers I'd be glad to hear your opinion on this.

**Comment by farindk** [comment]
Another option for handling this would be to forbid odd image sizes for 4:2:0 and use 4:4:4 whenever there are odd image sizes.

**Comment by fancycode** [comment]
For the record, I would reject unsupported combinations of image size / chroma, e.g. odd sizes with 4:2:0 with an error instead of silently changing the image size or the chroma format being used for encoding.

**Comment by carllee1993** [comment]
As you can see, the size of image is smaller than the stitched image when decoding the images comming from iphones. In my idea, it's similar to hvc1 and iden when decoding.

**Comment by carllee1993** [comment]
i have found the description in the ISO/IEC 23008-12:2017

**Box type: 'ispe'
Property type: Descriptive item property
Container: ItemPropertyContainerBox
Mandatory (per item): Yes
Quantity (per item): One
The ImageSpatialExtentsProperty documents the width and height of the associated image item. Every image item shall be associated with one property of this type, _prior to the association of all transformative properties_.**

**Comment by farindk** [comment]
Yes, and then ISO 23008 says for the fields in the ispe box:

    image_width specifies the width of the reconstructed image in pixels, as specified in 6.5.3.1.

and it says in 6.3:

    if the image item contains a coded image, the coded image is decoded and the reconstructed image is the output of the decoding process.

Hence, I understand this as `ispe size == coded image size`.

**Comment by GregoryJWolfe** [comment]
We have been using libheif/libheic for some time now, and just in the last few days received a customer's HEIC image with odd dimensions and YCbCr encoding.  This caused a crash of the module that was performing the image decoding.  So apparently IT IS possible to create an HEIC image with odd dimensions.

We are using version 1.0.0, but this problem looks like it is still present in the current GitHub code.  I fixed it in our copy of the source code by checking for odd dimensions, and then using the chrominance values from the last valid pixel/line when computing the last output pixel/line.

We (Kodak Alaris) would be happy to share our fix with the rest of the community.  Please contact me if you are interested.

Bottom line:  HEIC images with odd dimensions are out there, and your code needs to be ready for them.

**Comment by farindk** [comment]
This is supported already in libheif for quite some time. In images received by iOS, these are usually `grid` images with an odd grid size. It could also be done with a `clap` box, but iOS seems to be unable to read files with a `clap` transform.
